src/turbomind/models/llama/LlamaLinear.h

```diff
@@ -58,6 +69,41 @@ class LlamaLinear {
                               output_data,
                               weight.output_dims);
         sync_check_cuda_error();
+
+        if (lora_mask != nullptr && weight.lora_r > 0) {
+            cublas_wrapper_->Gemm(CUBLAS_OP_N,
+                                  CUBLAS_OP_N,
+                                  weight.lora_r,                                      // m
+                                  batch_size,                                         // n
+                                  weight.input_dims,                                  // k
+                                  (const T*)weight.lora_a,                            // A
+                                  weight.lora_r,                                      // lda
+                                  input_data,                                         // B
+                                  weight.input_dims,                                  // ldb
+                                  output_data + batch_size * weight.output_dims * 2,  // C
+                                  weight.lora_r);                                     // ldc
+
+            cublas_wrapper_->Gemm(CUBLAS_OP_N,
+                                  CUBLAS_OP_N,
+                                  weight.output_dims,                                 // m
+                                  batch_size,                                         // n
+                                  weight.lora_r,                                      // k
+                                  (const T*)weight.lora_b,                            // A
+                                  weight.output_dims,                                 // lda
+                                  output_data + batch_size * weight.output_dims * 2,  // B
+                                  weight.lora_r,                                      // ldb
+                                  output_data + batch_size * weight.output_dims,      // C
+                                  weight.output_dims);                                // ldc
+
+            invokeMaskAddTwoLinearOutput2(output_data,
+                                          output_data + batch_size * weight.output_dims,
+                                          weight.lora_scale,
+                                          lora_mask,
+                                          batch_size,
+                                          weight.output_dims,
+                                          stream_);
+            sync_check_cuda_error();
+        }
```

Comment:
This triples the size of all major intermediate buffers. From 600+MB to 2GB for 8K input tokens.

It can be computed as
```
output = lora(x)       // compute lora output first
output = mask(output)
output = x*W + output  // add gemm result to output by gemm API
```
because gemm computes C = alpha\*A\*x + beta*C. This way it will only require `bsz * r` additional workspace and the mask+add kernel can be simplified to mask-only kernel.